classes/user/Repository.inc.php

```diff
@@ -0,0 +1,333 @@
+    {
+        $workflowRoles = Application::get()->getWorkflowTypeRoles()[$workflowType];
+
+        if (array_key_exists($stageId, $userAccessibleStages)
+            && !empty(array_intersect($workflowRoles, $userAccessibleStages[$stageId]))) {
+            return true;
+        }
+        if (empty($userAccessibleStages) && in_array(Role::ROLE_ID_MANAGER, $userRoles)) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Check for roles that give access to the passed workflow stage.
+     *
+     * @param int $userId
+     * @param int $contextId
+     * @param Submission $submission
+     * @param int $stageId
+     *
+     * @return array
+     */
+    public function getAccessibleStageRoles($userId, $contextId, &$submission, $stageId)
+    {
+        $stageAssignmentDao = DAORegistry::getDAO('StageAssignmentDAO'); /** @var StageAssignmentDAO $stageAssignmentDao */
+        $stageAssignmentsResult = $stageAssignmentDao->getBySubmissionAndUserIdAndStageId($submission->getId(), $userId, $stageId);
+
+        $accessibleStageRoles = [];
+
+        // Assigned users have access based on their assignment
+        $userGroupDao = DAORegistry::getDAO('UserGroupDAO'); /** @var UserGroupDAO $userGroupDao */
+        while ($stageAssignment = $stageAssignmentsResult->next()) {
+            $userGroup = $userGroupDao->getById($stageAssignment->getUserGroupId(), $contextId);
+            $accessibleStageRoles[] = $userGroup->getRoleId();
+        }
+        $accessibleStageRoles = array_unique($accessibleStageRoles);
+
+        // If unassigned, only managers and admins have access
+        if (empty($accessibleStageRoles)) {
+            $roleDao = DAORegistry::getDAO('RoleDAO'); /** @var RoleDAO $roleDao */
+            $userRoles = $roleDao->getByUserId($userId, $contextId);
+            foreach ($userRoles as $userRole) {
+                if (in_array($userRole->getId(), [Role::ROLE_ID_SITE_ADMIN, Role::ROLE_ID_MANAGER])) {
+                    $accessibleStageRoles[] = $userRole->getId();
+                }
+            }
+            $accessibleStageRoles = array_unique($accessibleStageRoles);
+        }
+
+        return array_map('intval', $accessibleStageRoles);
+    }
+
+    /**
+     * Retrieves a filtered user report instance
+     *
+     *		@option int contextId Context ID (required)
+     *		@option int[] userGroupIds List of user groups (all groups by default)
+     */
+    public function getReport(array $args): Report
+    {
+        $dataSource = $this->getMany(
+            $this->getCollector()
+            ->filterByUserGroupids($args['userGroupIds'] ?? null)
+            ->filterByContextIds($args['contextId'] ?? [])
+        );
+        $report = new Report($dataSource);
+
+        HookRegistry::call('User::getReport', $report);
+
+        return $report;
+    }
+
+    public function getRolesOverview($args = [])
+    {
```

Comment:
This method should be passed a `Collector` with the required arguments rather than receiving them as an assoc array.

Reply:
:+1: Fixed!